File: efni/kennsluefni/talningarfraedi/code/fib.c
#include <stdio.h>
#define MOD 1000000009
typedef long long ll;

void multo(ll* a, ll* b, ll n)
{ // a *= b
    ll i, j, k, c[n][n];
    for (i = 0; i < n; i++) for (j = 0; j < n; j++) c[i][j] = 0;
    for (i = 0; i < n; i++) for (j = 0; j < n; j++) for (k = 0; k < n; k++)
        c[i][j] = (c[i][j] + (a[i*n + k]*b[k*n + j])%MOD)%MOD;
    for (i = 0; i < n; i++) for (j = 0; j < n; j++) a[i*n + j] = c[i][j];
}

void matpow(ll* a, ll p, ll n)
{ // a = a^p
    ll r[n][n], i, j;
    for (i = 0; i < n; i++) for (j = 0; j < n; j++) r[i][j] = 0;
    for (i = 0; i < n; i++) r[i][i] = 1;
    while (p > 0)
    {
        if (p%2 == 1) multo(*r, a, n);
        p /= 2;
        multo(a, a, n);
    }
    for (i = 0; i < n; i++) for (j = 0; j < n; j++) a[i*n + j] = r[i][j];
}

int fib(ll n)                                                                   // Reiknar n-tu Fibonacci töluna.
{
    ll a[2][2] = {{1, 1}, {1, 0}};                                              // Upphafstillum fylki fyrir rakningarvenslin.
    if (n == 1 || n == 2) return 1;                                             // Prentum upphafsgildin.
    matpow(*a, n - 2, 2);                                                       // Reiknum n - 2 veldið af flykinu.
    return (a[0][0] + a[0][1])%MOD;                                             // Margföldum efstu línuna í fylkinu við upphfsgildin og skilum.
}

int main()
{
    int n;
    scanf("%lld", &n);                                                          // Innlestur.
    printf("%lld\n", fib(n));                                                   // Reiknum og prentum svarið.
    return 0;
}

File: efni/kennsluefni/talningarfraedi/code/fib.java
import java.util.Scanner;

public class fib
{
    public static void main(String[] args)
    {
        Scanner s = new Scanner(System.in);
        System.out.println(fib(s.nextLong(), 1000000007));
    }
    
    static long fib(long n, long m)
    {
        long[][] a = {{1, 1}, {1, 0}};
        if (n <= 2) return 1;
        matpow(a, n - 2, m);
        return (a[0][0] + a[0][1])%m;
    }

    static void multo(long[][] a, long[][] b, long m)
    { // a *= b
        long[][] c = new long[a.length][a.length];
        for (int i = 0; i < a.length; i++) for (int j = 0; j < a.length; j++) for (int k = 0; k < a.length; k++)
            c[i][j] = (c[i][j] + (a[i][k]*b[k][j])%m)%m;
        for (int i = 0; i < a.length; i++) for (int j = 0; j < a.length; j++) a[i][j] = c[i][j];
    }

    static void matpow(long[][] a, long p, long m)
    { // a = a^p
        long[][] r = {{1, 0}, {0, 1}};
        while (p > 0)
        {
            if (p%2 == 1) multo(r, a, m);
            p /= 2;
            multo(a, a, m);
        }
        for (int i = 0; i < a.length; i++) for (int j = 0; j < a.length; j++) a[i][j] = r[i][j];
    }
}

